### PR TITLE
Separation of glyphs and characters for BitmapFontCache

### PR DESCRIPTION
**Pull request by NathanSweet**
This PR refactors how BitmapFontCache works. Previously it was assumed each character in a string mapped to a glyph for display. As discussed in #1589, this is not the case for some types of font rendering. This PR has two main goals: 1) to separate characters and glyphs, 2) to make parsing the string and preparing for display cleaner. The new approach works like this:

1. GlyphLayout, a new class, is given a string of characters. It first breaks the strings into "runs". A run is a string of characters all the same color and without newlines. The reason to break the text into runs is that each run can be "shaped" individually, which means to convert the characters into glyphs for display. When breaking the string into runs, GlyphLayout parses color markup tags and stores the color for each run.
1. Next, the runs are shaped using a new BitmapFontData method, `getGlyphs`. This takes the characters for the run and generates both a list of glyphs and a list of positions for each glyph. The default implementation just looks up the glyph for the character as has always been done. A more complex implementation could use [HarfBuzz](http://www.freedesktop.org/wiki/Software/HarfBuzz/) or another lib to do the shaping.
1. Next, the shaped runs are laid out. If wrapping is enabled and a run's glyphs exceed the wrap width, then glyphs are split for line wrapping using another new BitmapFontData method, `getWrapIndex`. The run is ended, the remaining glyphs are moved to a new run on the next line, and the layout process repeats. I'm not sure if line wrapping glyphs is the best solution, but at least where the wrap happens can be customized. Once all runs are laid out, if center or right alignment is requested then run positions are shifted appropriately.
1. At this point GlyphLayout is done. It stores the `Array<GlyphRun>` and provides a width and height which is the bounds of all the runs. Previously getting text bounds required all the work of laying out the glyphs, which then had to be repeated for drawing. With GlyphLayout, the work is only done once. Each GlyphRun has `Array<Glyph>` and a `FloatArray` for the positions.
1. Next comes BitmapFontCache. This class now has an `Array<GlyphLayout>` and its job is to store the vertex data for rendering the glyphs in those GlyphLayouts. Most of BitmapFontCache was left intact, eg it still can render from multiple texture regions.

There are a number of relatively small API breaking changes:

1. The BitmapFontCache `setText`, `setMultilineText`, `setWrappedText`, and similar `addXxx` methods are replaced by `setText` and `addText`. The number of overloads was kept reasonable. The same happened to the BitmapFont draw methods, which delegate to BitmapFontCache. These methods all return GlyphLayout instead of TextBounds.
1. BitmapFont.TextBounds is deprecated. I should have never added it and just used Vector2. Now, GlyphLayout has `width` and `height` and replaces TextBounds.
1. BitmapFont.HAlignment is gone. Align is used instead.
1. `markupEnabled` was moved to BitmapFontData.
1. BitmapFont `getBounds` and related methods were mostly removed, with the single most commonly used method deprecated. GlyphLayout should be used directly to avoid computing glyphs multiple times.
1. BitmapFont `computeGlyphAdvancesAndPositions` and `computeVisibleGlyphs` are deprecated and should be removed because they assume a 1:1 mapping of characters to glyphs. GlyphLayout can be used instead, which provides each glyph position. ~~I haven't made this change since it affects TextField and friends, but I will once this gets merged.~~ Now done.

Feedback on these changes is welcome. I've used the gdx-tests to verify things are still working. I also have updated Spine to use this, which is relatively complex real world usage. The font APIs are widely used so it's important we get this right. Now is a good time to break the API further if we need to.

### CODE REVIEW

**Comment by NathanSweet** [comment]
The more interesting files to look at:
[GlyphLayout](https://github.com/libgdx/libgdx/pull/2955/files#diff-62b695b6c37ee8ddcab0d32c9821f62d)
[BitmapFontCache](https://github.com/libgdx/libgdx/pull/2955/files#diff-a3a64db1ed704a93dede48b732d7aefb)
[BitmapFont](https://github.com/libgdx/libgdx/pull/2955/files#diff-e5b62781008351f68aa52151db58f0d6)

File: gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFont.java [hunk 1]
```diff
@@ -515,7 +362,7 @@ public int computeVisibleGlyphs (CharSequence str, int start, int end, float ava
 
 		for (; index < end; index++) {
 			char ch = str.charAt(index);
-			if (ch == '[' && markupEnabled) {
+			if (ch == '[' && data.markupEnabled) {
 				index++;
```
Comment (davebaol) [inline gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFont.java]: This is the issue mentioned by @MobiDevelop in https://github.com/libgdx/libgdx/issues/2886#issuecomment-83352300

Looks like this code has not changed drastically. So, if this PR is not merged soon I can fix and test the issue using the snapshot, then I can paste the fix here. Objections?

**Comment by davebaol** [comment]
@NathanSweet 
Hmmm.. your last commit says "Removed computeVisibleGlyphs".
Does this mean that we can consider issue #2886 fixed by this PR?

**Comment by NathanSweet** [comment]
I've now removed `computeVisibleGlyphs` and `computeGlyphAdvancesAndPositions`, replaced by GlyphLayout. Wasn't as bad as I thought it would be! This doesn't fix #2886 though, I will add a test and fix now.

You haven't looked into the code yet!?

**Comment by davebaol** [comment]
> You haven't looked into the code yet!?

Nope, I'm in office right now :smile:

**Comment by NathanSweet** [comment]
Offices typically have computers, yes? Get paid to libgdx! :wink:

File: gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java [hunk 2]
```diff
@@ -17,22 +17,32 @@
 package com.badlogic.gdx.graphics.g2d;
 
 import com.badlogic.gdx.graphics.Color;
-import com.badlogic.gdx.graphics.g2d.BitmapFont.BitmapFontData;
 import com.badlogic.gdx.graphics.g2d.BitmapFont.Glyph;
-import com.badlogic.gdx.graphics.g2d.BitmapFont.HAlignment;
-import com.badlogic.gdx.graphics.g2d.BitmapFont.TextBounds;
+import com.badlogic.gdx.graphics.g2d.GlyphLayout.GlyphRun;
+import com.badlogic.gdx.scenes.scene2d.utils.Align;
```
Comment (MobiDevelop) [inline gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java]: Perhaps Align has outgrown its place in the scene2d package. Though I guess we can't really move it now.

Comment (davebaol) [inline gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java]: Perhaps for backward compatibility you can leave it there as an empty class that simply extends the real `Align` class living out of the scene2d package.

Comment (NathanSweet) [inline gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java]: I've be ok with just moving it. Better to annoy everyone with a few breakages at once than to break them every few releases. Let's make a "break all yo shit" release! :hurtrealbad:

Comment (davebaol) [inline gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java]: Yeah I know that you dislike @Deprecated :smile: 
Anyways, it's ok with me. Break all the shit. :+1:

Comment (MobiDevelop) [inline gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFontCache.java]: Sounds good to me.

**Comment by NathanSweet** [comment]
Okie, I'll move Align.

Everybody ok with this PR? I'd like to merge so I can send another PR for on the fly FreeType rendering, good for CJK. If you need more time, just speak up!

**Comment by nooone** [comment]
Really nice PR!

> Let's make a "break all yo shit" release! :hurtrealbad:
BitmapFont still has a `@Deprecated getBounds(...)`  though. Maybe let's get rid of that as well then?

**Comment by NathanSweet** [comment]
I didn't want to, but you're right that it's better. BitmapFont `getBounds` and TextBounds are gone! I think the PR is ready!

**Comment by MobiDevelop** [comment]
I'm ok with merging this though, admittedly, I have not actually run it myself.

**Comment by NathanSweet** [comment]
That's the best kind of merge. ;)

Mario *says* he looked at some of the code. I expect he doesn't weigh in here so he can shout at me later.

Pressing the button!

**Comment by badlogic** [comment]
Fuck it, we do it live. BTW, I asked you to write a blog post so we can
people who don't read.

**Comment by NathanSweet** [comment]
Yes, of course my horse. With this in I can commit on the fly rendering, then have something to post about (today).

**Comment by nooone** [comment]
> I asked you to write a blog post so we can people who don't read.

What do we do with those people? I'm curious now...

**Comment by badlogic** [comment]
I let you find a verb :)

**Comment by davebaol** [comment]
Blaaaaame! What do I win? :P

**Comment by nooone** [comment]
You won the privilege of solving the next 3 issues that pop up!

**Comment by davebaol** [comment]
Cool, I'm going to add 3 bugs to the source and fix them immediately :)

**Comment by MobiDevelop** [comment]
@NathanSweet forgot to add GlyphLayout to the gdx.gwt.xml, @davebaol won the privilege of fixing that.

**Comment by davebaol** [comment]
Ok done! See commit https://github.com/libgdx/libgdx/commit/f415270b436fda63fa655a214fe72ee30963f4d5

-1, two missing :)

**Comment by MobiDevelop** [comment]
I have two more for you... These changes also broke the invaders and superjumper demos. :)

**Comment by davebaol** [comment]
@MobiDevelop 
Damn!  :/

@NathanSweet 
`BitmapFont.drawMultiLine()` has gone. What should I use instead?

**Comment by davebaol** [comment]
@NathanSweet 
BTW, these changes broke the Andorid MatrixTest
https://github.com/libgdx/libgdx/blob/master/tests/gdx-tests-android/src/com/badlogic/gdx/tests/android/MatrixTest.java#L78

**Comment by NathanSweet** [comment]
Oops. Silly GWT! :dancer: 

Silly demos, hiding in some other repo! Will look into it, unless davebaol is feeling frisky! ;) :D

@davebaol just use BitmapFont `draw`, they all do multiline (ie support newlines in the text). I'll fix up MatrixTest, sorry about that. Silly Android!

**Comment by NathanSweet** [comment]
MatrixTest fixed (forgot to link to this issue).

I added a code example section to the [blog post](http://www.badlogicgames.com/wordpress/?p=3658) to ease moving to the new API.

**Pull request merged**